### PR TITLE
Next.JS

### PR DESCRIPTION
**Pull request by tlgimenes**
## What's the purpose of this pull request?
This branch is an initial effort towards migrating base.store to Next.JS. Why should we migrate to Next.JS, you may ask. Below, I gathered the pros/cons of Base.store with Gatsby v4 vs Next.JS.

|     Dimension     | Next.JS | Gatsby v4 |
|:-----------------:|:-------:|:---------:|
|     Build Time    |   ~45s  |   ~100s   |
|      Dev Time     |   ~20s  |    ~38s   |
|  RAM usage (e2e)  |   84Mb  |   1.8Gb   |
|    Lighthouse     |    84   |     81    |
|       Stars       |  85.1k  |   52.7k   |
|        SSR        |   yes   |    yes    |
|        SSG        |   yes   |    yes    |
|        DSG        |   yes   |    yes    |
|        ISR        |   yes   |     no    |
|    CMS Preview    |   yes   |     no    |
|        i18n       |   yes   |     no    |
| Server Components |   yes   |     no    |
|      Plugins      |    no   |    yes    |

Also, another important aspect is developer choice. According to [npm trends](https://www.npmtrends.com/gatsby-vs-next), Next.JS was the chosen framework by developers
image

For the business user, Next.JS CMS Preview feature could help us building rich CMS Preview experiences, by allowing end users run on production bundles, instead of crappy dev bundles with HMR and other stuff.

Finally, Base.store on Next.JS runs in a wide variety of platforms, whereas Gatsby v4 version currently only runs on Webops.

## How does it work?
- Replaces Gatsby config files with Next ones
- Replaces Gatsby Link with Next Link component
- Replaces Gatsby data fetching API with Next ones

## How to test it?
Make sure nothing have changed.

## Checklist
- [x] CHANGELOG entry added

### CODE REVIEW

**Comment by igorbrasileiro** [comment]
@tlgimenes why not do it at https://github.com/vtex-sites/next.store ?

File: package.json [hunk 1]
```diff
@@ -34,33 +34,24 @@
...
-    "gatsby-plugin-gatsby-cloud": "^4.11.1",
-    "gatsby-plugin-manifest": "^4.11.1",
-    "gatsby-plugin-netlify": "^4.2.0",
-    "gatsby-plugin-next-seo": "^1.10.0",
-    "gatsby-plugin-nprogress": "^4.7.0",
-    "gatsby-plugin-robots-txt": "^1.7.0",
-    "gatsby-plugin-root-import": "^2.0.8",
-    "gatsby-plugin-sass": "^5.11.1",
-    "gatsby-plugin-webpack-bundle-analyser-v2": "^1.1.27",
+    "graphql": "15.x",
```
Comment (igorbrasileiro) [inline package.json:37]: Fix this version

**Comment by rafael-franca** [comment]
@tlgimenes How the prices are updated? Only using a useProduct query? I noted that not are using ISR, nor on-demand ISR.

File: src/components/ui/Image/Image.tsx [hunk 2]
```diff
@@ -1,32 +1,40 @@
 import { memo } from 'react'
-import { Helmet } from 'react-helmet-async'
+import Head from 'next/head'
```
Comment (igorbrasileiro) [inline src/components/ui/Image/Image.tsx:2]: Drop this component in favor of next/image

File: src/pages/[...slug].tsx [hunk 3]
```diff
@@ -1,74 +1,70 @@
...
-          ? maybeState.selectedFacets
-          : selectedFacets ?? [],
-      term: maybeState?.term ?? null,
-      sort: maybeState?.sort ?? 'score_desc',
+        state.selectedFacets.length > 0 ? state.selectedFacets : selectedFacets,
+    }
+  })
+
+  useEffect(() => {
+    const url = new URL(asPath, 'http://localhost')
```
Comment (igorbrasileiro) [inline src/pages/[...slug].tsx:42]: Hey, localhost here, is it right?

File: src/pages/_document.tsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { Head, Html, Main, NextScript } from 'next/document'
+
+import ThirdPartyScripts from 'src/components/ThirdPartyScripts'
+
+function Document() {
+  return (
+    <Html>
+      <Head>
+        <ThirdPartyScripts />
```
Comment (igorbrasileiro) [inline src/pages/_document.tsx:9]: We can move this to the _app.ts component and wrap it with next/head

File: src/pages/s.tsx [hunk 5]
```diff
@@ -1,45 +1,36 @@
...
+
+import storeConfig from '../../store.config'
+
+const useSearchParams = () => {
   const [params, setParams] = useState<SearchState | null>(null)
+  const { asPath } = useRouter()
 
   useEffect(() => {
-    const url = new URL(href)
+    const url = new URL(asPath, 'http://localhost')
```
Comment (igorbrasileiro) [inline src/pages/s.tsx:21]: also here

File: vtex.env [hunk 6]
```diff
@@ -1,12 +1,7 @@
...
+NEXT_TELEMETRY_DISABLED=true
 
 # configures WebOps
 # Set to false while the WebOps team fixes the Jira's DXWP-279 issue.
 USE_NODE_MODULES_CACHE=false
-USE_GATSBY_CACHE=true
-
-# 🐞🐞 Uncomment for debugging final bundle 🐞🐞
-# GATSBY_STORE_PROFILING=true
+USE_GATSBY_CACHE=false
```
Comment (igorbrasileiro) [inline vtex.env:7]: drop it.

**Review by igorbrasileiro — CHANGES_REQUESTED** [review]
1. Why not use SWC instead of babel?
2. Why not use next/image instead own component
3. Why not use next server components https://nextjs.org/docs/advanced-features/react-18/server-components from react 18?
4. Drop _document.js and add the thirdpartyscripts inside the _app.ts

**Comment by tlgimenes** [comment]
> 1. Why not use SWC instead of babel?
> 2. Why not use next/image instead own component
> 3. Why not use next server components https://nextjs.org/docs/advanced-features/react-18/server-components from react 18?
> 4. Drop _document.js and add the thirdpartyscripts inside the _app.ts

We will do this eventually. Le'ts keep the migration to a minimal change.

File: src/components/ui/Image/Image.tsx (hunk 2 above)
Comment (tlgimenes) [reply src/components/ui/Image/Image.tsx:2]: we can do this in a later PR

File: src/pages/[...slug].tsx (hunk 3 above)
Comment (tlgimenes) [reply src/pages/[...slug].tsx:42]: yes, we don't use the origin

File: public/~partytown/debug/partytown-atomics.js [hunk 7]
```diff
@@ -0,0 +1,526 @@
+/* Partytown 0.5.2 - MIT builder.io */
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply public/~partytown/debug/partytown-atomics.js:1]: nice!

File: src/components/sections/ProductGallery/ProductGallery.tsx [hunk 8]
```diff
@@ -26,12 +27,11 @@ interface Props {
...
 function ProductGallery({ title, searchTerm }: Props) {
   const [isFilterOpen, setIsFilterOpen] = useState<boolean>(false)
-  const { pages, addNextPage, addPrevPage, state: searchState } = useSearch()
-
+  const { pages, addNextPage, addPrevPage } = useSearch()
   const { data } = useGalleryQuery()
   const facets = useDelayedFacets(data)
   const totalCount = data?.search.products.pageInfo.totalCount ?? 0
-  const { next, prev } = usePagination(totalCount)
+  const { next, prev } = useDelayedPagination(totalCount)
```
[the remark this answers is not in the document]
Comment (tlgimenes) [reply src/components/sections/ProductGallery/ProductGallery.tsx:34]: no, we can do it differently later on. Let's keep these changes to a minimum

File: src/pages/_document.tsx (hunk 4 above)
Comment (tlgimenes) [reply src/pages/_document.tsx:9]: doing this will include more JS on the final bundle than needed

Comment (igorbrasileiro) [reply src/pages/_document.tsx:9]: Why? Since the next/head is already in use?

Comment (igorbrasileiro) [reply src/pages/_document.tsx:9]: https://partytown.builder.io/nextjs#configure

File: src/pages/s.tsx (hunk 5 above)
Comment (igorbrasileiro) [reply src/pages/s.tsx:21]: Can explain better?

**Comment by irgmedeiros** [comment]
Screen Shot 2022-04-20 at 18 28 37

What do plugins means here?

File: .eslintrc [hunk 9]
```diff
@@ -1,14 +1,21 @@
 {
-  "extends": "vtex-react/gatsby",
+  "extends": [
+    "vtex-react/gatsby",
```
Comment (filipewl) [inline .eslintrc:3]: Can we also remove this?

File: .eslintrc [hunk 10]
```diff
@@ -1,14 +1,21 @@
...
+    "next/core-web-vitals"
+  ],
   "root": true,
   "rules": {
     "react/prop-types": "off",
-    "react/react-in-jsx-scope": "off"
+    "react/react-in-jsx-scope": "off",
+    "jsx-a11y/anchor-is-valid": "off"
```
[the remark this answers is not in the document]
Comment (filipewl) [reply .eslintrc:10]: Found some explanation here: https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-is-valid.md#case-i-use-nextjs-and-im-getting-this-error-inside-of-links

File: CHANGELOG.md [hunk 11]
```diff
@@ -8,10 +8,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [Unreleased]
 
 ### Added
-
 - Applies new local tokens to `Badge` (#462)
 - Applies new local tokens to `Hero` (#435)
 - Applies new local tokens to `Quantity Selector` (#448)
+
+### Changed
+- Migrates to Next.JS (#475)
```
Comment (filipewl) [inline CHANGELOG.md:16]:
```suggestion
- Migrates from Gatsby v4 to Next.JS (#475)
```

File: src/pages/[...slug].tsx [hunk 12]
```diff
@@ -1,76 +1,70 @@
-import { parseSearchState, SearchProvider, useSession } from '@faststore/sdk'
```
Comment (filipewl) [inline src/pages/[...slug].tsx:1]: Noticed when clicking the nav links that the products here weren't getting updated. A full page refresh was required to update it.

https://user-images.githubusercontent.com/381395/164520813-0f4f8aaa-fa35-49f4-af3c-29858a4a12a4.mp4

Comment (tlgimenes) [reply src/pages/[...slug].tsx:1]: nice catch! fixed!

File: .eslintrc (hunk 9 above)
Comment (tlgimenes) [reply .eslintrc:3]: not really. We need to have our own linting rules :(

File: src/pages/_document.tsx (hunk 4 above)
Comment (igorbrasileiro) [reply src/pages/_document.tsx:9]: Take a look here using the[ partytown inside the _app with next/head](https://github.com/vtex-sites/next.store/pull/8): the view source view-source:https://next-store-l8cbr9fj1-faststore.vercel.app/

Your branch view-source: view-source:https://sfj-b168bce--base.preview.vtex.app/

Comment (tlgimenes) [reply src/pages/_document.tsx:9]: This is the report with the <ThirdPartyScripts/> in _app vs on __document:
on _app:
```
Page                                       Size     First Load JS
┌ ○ /                                      2.62 kB         109 kB
├   /_app                                  0 B            97.3 kB
├ ● /[...slug]                             2.21 kB         113 kB
├ ● /[slug]/p                              4.56 kB         111 kB
├ ○ /404                                   431 B           103 kB
├ ○ /500                                   439 B           103 kB
├ ○ /account                               368 B           103 kB
├ λ /api/graphql                           0 B            97.3 kB
├ ○ /checkout                              374 B           103 kB
├ ○ /login                                 368 B           103 kB
└ ○ /s                                     2.15 kB         109 kB
+ First Load JS shared by all              97.3 kB
  ├ chunks/framework-fc97f3f1282ce3ed.js   44.9 kB
  ├ chunks/main-4cf47fefcd9ebea2.js        20.5 kB
  ├ chunks/pages/_app-7bc8dbb918c461c6.js  30.5 kB
  ├ chunks/webpack-566f5c6c8d77736b.js     1.55 kB
  └ css/7d73de50030d3a43.css               14.7 kB
```

on __document:
Page                                       Size     First Load JS
┌ ○ /                                      2.62 kB         110 kB
├   /_app                                  0 B             100 kB
├ ● /[...slug]                             2.21 kB         114 kB
├ ● /[slug]/p                              4.56 kB         112 kB
├ ○ /404                                   431 B           104 kB
├ ○ /500                                   439 B           104 kB
├ ○ /account                               368 B           104 kB
├ λ /api/graphql                           0 B             100 kB
├ ○ /checkout                              374 B           104 kB
├ ○ /login                                 368 B           104 kB
└ ○ /s                                     2.15 kB         110 kB
+ First Load JS shared by all              100 kB
  ├ chunks/framework-fc97f3f1282ce3ed.js   44.9 kB
  ├ chunks/main-4cf47fefcd9ebea2.js        20.5 kB
  ├ chunks/pages/_app-6d4dd2119ba131d9.js  33.2 kB
  ├ chunks/webpack-786ba62b2546136e.js     1.55 kB
  └ css/7d73de50030d3a43.css               14.7 kB

Comment (igorbrasileiro) [reply src/pages/_document.tsx:9]: Ok, thank you for the tests

**Comment by igorbrasileiro** [comment]
> Screen Shot 2022-04-20 at 18 28 37
> 
> What do plugins means here? A way to extend the sites using the framework? As far as I know both are extensible although the experience on Gatsby seems to be way easier. Also I don't know yet if the plugins in NextJs are limited somehow.
> 
> Some references on extending NextJs with plugins:
> 
> https://www.grouparoo.com/blog/nextjs-plugins https://github.com/cyrilwanner/next-compose-plugins

Hey @irgmedeiros , IMO, we see plugins as a negative thing, because we're tight coupled with the framework. In NextJS world, we don't need it, we use npm packages or whatever js code.

**Comment by tlgimenes** [comment]
now available at https://github.com/vtex-sites/nextjs.store

**Pull request closed without merging**